### PR TITLE
Fix scroll related issues

### LINKED ISSUE #7017: Fix handling of hash links & scrolling
## Description

Currently Gatsby handles clicking on hash links as in we do correctly scroll to the link. However, we don't handle scrolling back to the previous scroll position when you click back.

### Steps to reproduce

- navigate to https://next.gatsbyjs.org/docs/performance-tracing/
- click on "Tracing backend examples"
- click back button
- observe nothing happens

### Expected result

When clicking the back button, it'd scroll to the previous scroll position.

### Actual result

Nothing happens.

## How to fix

After https://github.com/gatsbyjs/gatsby/pull/6918 lands, we should investigate adding remembering the hash scroll position in either gatsby-react-router-scroll or in the new top-level `RouteHandler` component.

### LINKED ISSUE #7921: [using-page-transitions] Scrolls to top before animating
On the `using-page-transitions` example, the page scrolls to top just before fade-out animations starts.
If you have a long page, and click a link to trigger a route change, it immediately scrolls to top of the current page, then the fade-out animation begins, then the new page loads in.
I see the `gatsby-react-router-scroll` is a gatsby dependency - is there a way to pass a timeout so you don't get that page flash?

This behaviour can be checked here https://www.kaizen.co.uk/

### PR DESCRIPTION
**Pull request by stefanprobst**
Attempt to solve some scroll related issues:
* remove hash-link scroll handling from `<Link />` and put it in `shouldUpdateScroll`. Also, let `scroll-behavior` handle scrolling to element ids. This should fix #7017, #7450, #8019.
* make the `shouldUpdateScroll` browser API more flexible by passing the whole routerProps object, and allowing access to the saved scroll positions. This allows things like delaying scroll updates until after a page transition animation has finished, should fix #7921. Maybe this one is a little late for v2 and should be separated into another PR.

### CODE REVIEW

File: packages/gatsby/cache-dir/navigation.js [hunk 1]
```diff
@@ -97,10 +97,12 @@ const navigate = (to, options = {}) => {
   })
 }
 
-function shouldUpdateScroll(prevRouterProps, { location: { pathname } }) {
+function shouldUpdateScroll(prevRouterProps, routerProps) {
+  const { location: { pathname, hash } } = routerProps
   const results = apiRunner(`shouldUpdateScroll`, {
     prevRouterProps,
-    pathname,
```
Comment (pieh) [inline packages/gatsby/cache-dir/navigation.js]: I think that if we leave `pathname` here for backward compatibility, this change would potentially could go in

**Comment by stefanprobst** [comment]
This should be good to go, let me know if there's anything else missing from my side.

**Comment by joltmode** [comment]
You should update the docs @ https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/utils/api-browser-docs.js#L56

File: packages/gatsby/cache-dir/navigation.js [hunk 2]
```diff
@@ -97,10 +97,16 @@ const navigate = (to, options = {}) => {
...
+function shouldUpdateScroll(prevRouterProps, routerProps) {
+  const {
+    location: { pathname, hash },
+  } = routerProps
   const results = apiRunner(`shouldUpdateScroll`, {
     prevRouterProps,
+    // `pathname` for backwards compatibility
     pathname,
+    routerProps,
+    savedScrollPositions: this._stateStorage,
```
Comment (joltmode) [inline packages/gatsby/cache-dir/navigation.js]: The naming here doesn't match the actual behavior. You don't get positions, you only get the storage.

File: packages/gatsby/cache-dir/navigation.js [hunk 3]
```diff
@@ -97,10 +97,16 @@ const navigate = (to, options = {}) => {
...
-function shouldUpdateScroll(prevRouterProps, { location: { pathname } }) {
+function shouldUpdateScroll(prevRouterProps, routerProps) {
+  const {
+    location: { pathname, hash },
+  } = routerProps
   const results = apiRunner(`shouldUpdateScroll`, {
     prevRouterProps,
+    // `pathname` for backwards compatibility
     pathname,
+    routerProps,
```
Comment (joltmode) [inline packages/gatsby/cache-dir/navigation.js]: I'm scared about `history` being passed here, same as I'm not a fan of `location.replace`/`location.reload`. 

I wouldn't expose functionality that could be abused outside the normal loop of things and potentially causing unexpected results.

File: packages/gatsby/cache-dir/navigation.js (hunk 2 above)
Comment (stefanprobst) [reply packages/gatsby/cache-dir/navigation.js]: Yes, will update the docs with an example. Maybe it's enough to expose `getSavedScrollPosition: this._stateStorage.read`?

Comment (joltmode) [reply packages/gatsby/cache-dir/navigation.js]: That'd seem to be better, yeah.

**Comment by stefanprobst** [comment]
Updated the docs, plucked off `location` from `routerProps`, and `read()` from the scroll position session store.
Still to do: docs should say what the return value of `shouldUpdateScroll` means, and maybe also that the positions are internally stored in SessionStorage, so survive a reload.
Question: Passing on `read()` works because of [this, which is marked as a "hack"](https://github.com/gatsbyjs/gatsby/blob/beb669f43e4d587b516cf291a3cfeedf0817e30e/packages/gatsby-react-router-scroll/src/ScrollBehaviorContext.js#L78). Is it okay to rely on that?

**Comment by joltmode** [comment]
> Updated the docs, plucked off `location` from `routerProps`, and `read()` from the scroll position session store.

`history`, not `location`. :slightly_smiling_face:

**Comment by stefanprobst** [comment]
Help! Could anybody give me some hints how to turn this into proper jsdoc, thanks!
```js
/**
 * Allow a plugin to decide if the scroll position should update or
 * not on a route change.
 * 
 * @callback getSavedScrollPosition
 * @param {object} location A location object.
 * @param {string} location.pathname The path of the route.
 * @returns {Array|null} An [x, y] array of coordinates, or `null` if nothing was found.
 *
 * @function shouldUpdateScroll
 * @param {object} $0
 * @param {object} $0.prevRouterProps The previous state of the router before the route change.
 * @param {object} $0.routerProps The current state of the router.
 * @param {string} $0.pathname The new pathname (for backwards compatibility with v1).
 * @param {getSavedScrollPosition} $0.getSavedScrollPosition Takes a location and returns the coordinates of
 * the last scroll position for that location, or `null`. Gatsby saves scroll positions for each route
 * in `SessionStorage`, so they are available after page reload.
 * @returns {boolean|string|Array} Should return either an [x, y] Array of coordinates to scroll to,
 * a string of the `id` or `name` of an element to scroll to, `false` to not update the scroll position,
 * or `true` for the default behavior.
 * @example
 * exports.shouldUpdateScroll = ({
 *   routerProps: { location },
 *   getSavedScrollPosition
 * }) => {
 *   const currentPosition = getSavedScrollPosition(location)
 *   const queriedPosition = getSavedScrollPosition({ pathname: '/random' })
 *
 *   window.scrollTo(...(currentPosition || [0, 0]))
 *
 *   return false
 * }
 */
exports.shouldUpdateScroll = true
```

**Comment by joltmode** [comment]
@stefanprobst what exactly is the problem over there?

**Comment by stefanprobst** [comment]
@joltmode Neither `@callback` nor the `@returns` show up (nothing in React DevTools). This is all very probably because I'm doing this wrong, but could also be something to do with `documentationjs` or the `transformer` gatsby uses.

Unfortunately I could not find an example in the gatsby docs to emulate - the closest seems `onCreateWebpackConfig.getConfig`, which incidentally is also broken.

So my question is how to document `getSavedScrollPosition` and the return values correctly?

**Comment by stefanprobst** [comment]
Updated the docs. Should be good now, apart from the fact that the return values don't show on the website, but that's a bug in `gatsby-transformer-documentationjs` (does not work for other documented APIs either).

**Pull request merged**